### ISSUE TITLE
Add explanation about thin-edge.io role in firmware updates

### ISSUE THREAD

**Issue by reubenmiller** [thin-edge/thin-edge.io#2733]
**Is your documentation request related to a problem? Please describe.**

thin-edge.io's role in the firmware update process is currently missing from the docs.

**Describe the improvement you'd like**

Add a section or page to explain how thin-edge.io approaches firmware updates:
* interface with firmware update tools (such as rauc, rugpi, SWUpdate?, mender etc.)
* Validate the device's connectivity after an A/B partition swap
* Role in being included in the image building process etc.

**Additional context**

**Comment by geotom** [comment]
I would also love to see more documentation of how firmware updates in thin-edge are done and what possible scenario are supported. The old architecture image features Mender, but I have nowhere ever found any description if this is already supported.

- Does this work already?
- Does it only work with C8?
- What Firmware providers can be integrated?

My feeling is that this feature was never really implemented, or am I missing something.
Thanks for an answer or updated documents.

**Comment by reubenmiller** [comment]
> I would also love to see more documentation of how firmware updates in thin-edge are done and what possible scenario are supported. The old architecture image features Mender, but I have nowhere ever found any description if this is already supported.
> 
> * Does this work already?
> * Does it only work with C8?
> * What Firmware providers can be integrated?
> 
> My feeling is that this feature was never really implemented, or am I missing something. Thanks for an answer or updated documents.

Yes the firmware update feature is very much implemented, but it is currently only documented in the [Extend > Firmware Management](https://thin-edge.github.io/thin-edge.io/extend/firmware-management/) section, and does not go into the full technical details on how it works, though it does give the references to the Yocto layers which include the integration with different update technology such as Mender, Rauc or Rugpi (though we're currently doing some refactoring on the integration now that Rugpi has been renamed to Rugix).

But in short, this is very complex topic to describe to users who don't have any experience in doing A/B OS updates (what we call firmware updates), as the A/B update technology has a lot of pre-requisites like selecting bootloader, partition layout, and the A/B application (e.g. mender/rauc) must already be installed in the initial image etc...so you generally can't roll this out to brownfield devices without having to re-flash the device.

From a thin-edge.io perspective, there isn't much "firmware specific" functionality that had to be implemented (apart from translating some firmware update messages to Cumulocity). The firmware update mechanism is solely enabled by the [tedge-agent workflow engine](https://thin-edge.github.io/thin-edge.io/references/agent/operation-workflow/).  The workflow is a user-defined state machine that is trigger upon receiving the user specified operation (`firmware_update` in this case). So for any integration with the A/B update tech (e.g. Mender, Rauc, Rugix), the tie-in point with thin-edge.io is just to create a `firmware_update.toml` workflow to execute the steps as mandated by the A/B update tech that you're using (as each tool has different commands, and slightly different ways of doing things). The developer still has to build the image and update bundles,  though we do have Yocto layers, and Rugix recipes available for this, but some of the integration is dependent on the specific board that you are integrating with (e.g. sometimes custom u-boot scripts have to be created etc., but this had more to do with the A/B update tech rather than thin-edge.io directly).


Regardless what A/B update tech is being used, the integration usually involves the following steps:

* Create a firmware_update.toml workflow (we already have Rauc, Mender and Rugix workflows available)
* Build a device image which includes the A/B update tool and all of its pre-requisites to be used for the initial flashing of the device
* Build an update bundle which is then used for Over-the-Air updates

**Comment by geotom** [comment]
Thanks for the fast response! I have already worked with mender and was curious how this is integrated with thinedge. Mender alone could be used for pushing updates, but for me it was not clear how this is supported here.

Where do I find the firmware_update.toml for the already supported Mender, Rauc, Rugix?

I think the documentation for this should be enhanced. Because right now I was quite disappointed to read there only about how certain images are created (like for Raspberry, etc) but no info about the workflow and what is supported, etc. And an example would greatly improve the docs as well.

**Comment by reubenmiller** [comment]
> Thanks for the fast response! I have already worked with mender and was curious how this is integrated with thinedge. Mender alone could be used for pushing updates, but for me it was not clear how this is supported here.

The mender integration is using mender in standalone mode (which is the open-source component of Mender). In the end thin-edge.io is the glue between the cloud and mender. It receives the firmware update operation from the cloud, translates the URL to something that can be used directly by the mender client (for Cumulocity, the URL will be translated to a local URL which uses the Cumulocity local proxy service), then the workflow will make the mender client cli calls, restart the device, and then call the `mender commit` etc.

Below are some links within the meta-tedge Yocto layer which includes both the recipes that are used to install thin-edge.io and also configure the firmware update related parts (for Mender and RAUC at least), though the links point to the workflows for each technology:

* [thin-edge.io Mender workflow](https://github.com/thin-edge/meta-tedge/blob/kirkstone/meta-tedge-mender/recipes-tedge/tedge-firmware/tedge-firmware/firmware_update.mender.toml)
* [thin-edge.io RAUC workflow](https://github.com/thin-edge/meta-tedge/blob/kirkstone/meta-tedge-rauc/recipes-tedge/tedge-firmware-rauc/tedge-firmware-rauc/firmware_update.rauc.toml)

> I think the documentation for this should be enhanced. Because right now I was quite disappointed to read there only about how certain images are created (like for Raspberry, etc) but no info about the workflow and what is supported, etc. And an example would greatly improve the docs as well.

Yes, we could mention the different types of projects that can be built (e.g. we have both Mender and RAUC layers). If you're using Yocto, you can build for different machines (Machine is Yocto terminology), however each machine will generally have a different BSP (Board Support Package) which might need some tweaking to be compatible with the different firmware update technologies (e.g. sometimes this requires writing a u-boot script to match your actual device)...but all of these components are part of a device integration, and not directly related to thin-edge.io, but some users will touch on these topics for the first time because they start using thin-edge.io (so it is sometimes difficult to communicate this difference).

**Comment by reubenmiller** [comment]
@geotom But if you're interested in a more detailed discussion about firmware A/B updates and doing a device integration for a specific board, then there are also people available from professional services that can help assist (this is a paid service).

**Comment by geotom** [comment]
> [@geotom](https://github.com/geotom) But if you're interested in a more detailed discussion about firmware A/B updates and doing a device integration for a specific board, then there are also people available from professional services that can help assist (this is a paid service).

Thanks. I think I will try myself to get this working. But thanks for linking to the relevant files. So if I understand correctly, the firmware update is only available & working if I use a Yocto build using [meta-tedge](https://github.com/thin-edge/meta-tedge/tree/kirkstone).

I was looking at [mender-mcu](https://github.com/mendersoftware/mender-mcu) for ESP32 devices. As thinedge does not run on ESP32, I think the only way to get this working together could be the child device protocol (triggerering the mender-mcu via MQTT topics on an ESP32 that works as a child device). Do child devices support firmware update commands?

**Comment by reubenmiller** [comment]
> Thanks [@reubenmiller](https://github.com/reubenmiller). I think I will try myself to get this working. But thanks for linking to the relevant files. So if I understand correctly, the firmware update is only available & working if I use a Yocto build using [meta-tedge](https://github.com/thin-edge/meta-tedge/tree/kirkstone).

Yes, though specific examples are using Yocto (which is the industry standard when it comes to embedded linux devices). You could do something similar with Buildroot (an alternative to Yocto), but we don't have an official repo for Builtroot yet.

You could also use a more begineer-friendly tool called [Rugix](https://github.com/silitics/rugix), but we're still updating the thin-edge.io [Rugix integration repository](https://github.com/thin-edge/tedge-rugpi-core), but hopefully we should be done in the next few weeks.

But all of the forementioned technologies require you to build a full OS image yourself, as each integration needs fine-grain control over the image in order to do the partition swapping during an update in a reliable manner.

> I was looking at [mender-mcu](https://github.com/mendersoftware/mender-mcu) for ESP32 devices. As thinedge does not run on ESP32, I think the only way to get this working together could be the child device protocol (triggerering the mender-mcu via MQTT topics on an ESP32 that works as a child device). Do child devices support firmware update commands?

Yes, you can write your custom child device agent which runs on a microcontroller, and if it tells thin-edge.io that it supports the `firmware_update`command, then it can be used. You don't specifically need mender for this though, as the ESP32 do support some firmware updates out of the box (if you're using the espressif sdk). We have an example of an ESP32 client, https://github.com/thin-edge/freertos-esp32-client, which connects to a local thin-edge.io instance and can then handle firmware_update operation (which under the hood just calls the required espressif API). Whilst thin-edge.io can't run on a microcontroller, having thin-edge.io running somewhere else in the local network (e.g. on a gateway device), significantly simplifies the microcontroller development...and microcontrollers generally have very constrained resources anyway, so half of the thin-edge.io functionality does not make sense to run on a microcontroller (e.g. the are pretty much useless as a server function as they can't server many clients concurrently).

**Comment by geotom** [comment]
I'm familiar with Yocto, so good to know it is available via the meta-thinedge  layer. And next to ESP32, we will also use other devices, so having one OTA platform like Mender for all of them makes more sense than using individual ones. I saw the FreeRTOS extension and think we will go down a similar route. Maybe it could be an idea for thinedge, to provide microcontroller-based agent implementations, implementing only a subset of the functionalities. Thanks for all the info and confirming my assumptions [@reubenmiller](https://github.com/reubenmiller)!

**Comment by reubenmiller** [comment]
Though we still believe that the open nature of the local MQTT interface should be relatively easy for users to implement using their own preferred libraries and language. The microcontroller world is very special and often requires a lot of tricks and workarounds, and generally it is more akin to an SDK rather than an application.

But we’re always more than happy to explore the difference spaces out there and collaborate if people are willing.
